File: clients/OpenSync.Client.Js/src/SyncDocument.ts
import { OpenSyncClient, SyncEvent } from './OpenSyncClient';

export class SyncDocument {
  private client: OpenSyncClient;
  private name: string;

  constructor(client: OpenSyncClient, name: string) {
    this.client = client;
    this.name = name;
  }

  async subscribe(): Promise<void> {
    await this.client.subscribe('document', this.name);
  }

  async set(data: any, expectedRevision?: number): Promise<void> {
    await this.client.httpPost(`/documents/${this.name}`, { data, expected_revision: expectedRevision });
  }

  onUpdated(handler: (data: any) => void): void {
    this.client.on('object_updated', (event: SyncEvent) => {
      if (event.object_id === this.name) handler(event.data);
    });
  }

  onRemoved(handler: () => void): void {
    this.client.on('object_removed', (event: SyncEvent) => {
      if (event.object_id === this.name) handler();
    });
  }
}

File: nupkgs/OpenSync.Sdk/SyncDocument.cs
using OpenSync.Sdk.Events;

namespace OpenSync.Sdk;

public class SyncDocument
{
    private readonly OpenSyncClient _client;
    private readonly string _uniqueName;

    public event EventHandler<ItemChangedEventArgs>? OnUpdated;
    public event EventHandler<ItemChangedEventArgs>? OnRemoved;

    public SyncDocument(OpenSyncClient client, string uniqueName)
    {
        _client = client;
        _uniqueName = uniqueName;
        _client.OnEvent += HandleEvent;
    }

    private void HandleEvent(object? sender, SyncEventArgs args)
    {
        if (args.ObjectType == "document" && args.ObjectId == _uniqueName)
        {
            if (args.Event == "object_updated")
                OnUpdated?.Invoke(this, new ItemChangedEventArgs(args.Data));
            else if (args.Event == "object_removed")
                OnRemoved?.Invoke(this, new ItemChangedEventArgs(args.Data));
        }
    }

    public async Task SetAsync(object data, long? expectedRevision = null)
    {
        var json = System.Text.Json.JsonSerializer.Serialize(data);
        await _client.HttpPostAsync<object>($"/documents/{_uniqueName}", new { data = json, expected_revision = expectedRevision });
    }

    public async Task SubscribeAsync() => await _client.SubscribeAsync("document", _uniqueName);
}
